libcontainer/process_linux.go

```diff
@@ -321,13 +321,18 @@ func (p *setnsProcess) execSetns() error {
 // terminate sends a SIGKILL to the forked process for the setns routine then waits to
 // avoid the process becoming a zombie.
 func (p *setnsProcess) terminate() error {
-	if p.cmd.Process == nil {
-		return nil
+	var err error
+	if p.cmd.Process != nil {
+		err = p.cmd.Process.Kill()
+		if _, werr := p.wait(); err == nil {
+			err = werr
+		}
 	}
-	err := p.cmd.Process.Kill()
-	if _, werr := p.wait(); err == nil {
+
+	if _, werr := unix.Wait4(-1, nil, 0, nil); werr != nil {
```

Comment:
Now that I think about it, I don't think we should to do this here -- we do need to add a `wait` in stage-0 instead. At the moment we send `stage1_pid` over the pipe, but instead we can add the wait after we finish the "stage-1 synchronisation loop" and remove the `stage1_pid` JSON logic entirely.

Using `PR_SET_CHILD_SUBREAPER` can lead to some other issues (other users of libcontainer that have other children will have unexpected reparenting behaviour, and if we do a wait here we might wait for a different child than we wanted).

(Also if we're waiting for more than one process you need to do wait in a loop -- and if you want to be sure you got all of them you need to do it until `-ECHILD`. But there might be other children so we can't do this and it would lead to all sorts of other issues...)

Reply:
> we do need to add a `wait` in stage-0 instead.

There are two problems:
1. If the `stage-0` suddenly dead, how to reap `stage-1` and `stage-2`?
2. For `runc exec -t`, the `stage-0` process has exited after the container started, if we wait it in `stage-0`, runc process will  have no way to know whether the container has exited or not.

Reply:
> Also if we're waiting for more than one process you need to do wait in a loop -- and if you want to be sure you got all of them you need to do it until `-ECHILD`.

Yes, thanks.

> try_mapping_tool already does a waitpid.

Yes, but if `stage-0` process suddenly dead, I think runc should reap all these process.
As we all know, `PR_SET_CHILD_SUBREAPER` means that runc has the ability to reap grand child processes, but if the child process has not exited, runc will not reap it's grand child process.

I think the only problem is that we should tell the `libct/nsexec` users to add `unix.Prctl(unix.PR_SET_CHILD_SUBREAPER, 1, 0, 0, 0)` in their system like in runc.

Reply:
https://grep.app/search?q=github.com/opencontainers/runc/libcontainer/nsenter&filter[lang][0]=Go
It seems that there a 5 public repos using `libct/nsenter`.